File: backend/routers/system_admin.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from sqlalchemy import text
from datetime import datetime, timedelta
from typing import List, Dict, Any, Optional
import psutil
import redis
import os
import subprocess
import json
from database import get_db
from auth import get_current_user
import models

router = APIRouter(prefix="/admin", tags=["System Administration"])

# Mock data for development - replace with real implementations
def get_mock_system_health():
    """Mock system health data"""
    return {
        "overall": {
            "status": "healthy",
            "score": 85,
            "message": "System is operating normally"
        },
        "services": [
            {
                "name": "backend",
                "status": "healthy",
                "uptime": "2d 14h 32m",
                "cpu": 15.2,
                "memory": 45.8,
                "lastRestart": datetime.now() - timedelta(days=2, hours=14)
            },
            {
                "name": "database",
                "status": "healthy", 
                "uptime": "7d 3h 15m",
                "cpu": 8.5,
                "memory": 62.3,
                "lastRestart": datetime.now() - timedelta(days=7, hours=3)
            },
            {
                "name": "redis",
                "status": "healthy",
                "uptime": "5d 18h 42m", 
                "cpu": 3.2,
                "memory": 28.7,
                "lastRestart": datetime.now() - timedelta(days=5, hours=18)
            }
        ],
        "resources": {
            "cpu": {
                "current": 25.4,
                "average": 22.1,
                "trend": "stable"
            },
            "memory": {
                "used": 4294967296,  # 4GB
                "total": 8589934592,  # 8GB
                "percentage": 50.0,
                "trend": "stable"
            },
            "disk": {
                "used": 107374182400,  # 100GB
                "total": 214748364800,  # 200GB
                "percentage": 50.0,
                "trend": "up"
            },
            "network": {
                "inbound": 1024000,
                "outbound": 512000
            }
        },
        "security": {
            "sslCertificate": {
                "domain": "localhost",
                "issuer": "Self-Signed",
                "validFrom": datetime.now() - timedelta(days=30),
                "validTo": datetime.now() + timedelta(days=335),
                "daysUntilExpiry": 335,
                "status": "valid",
                "autoRenewal": True
            },
            "securityHeaders": {
                "hsts": True,
                "csp": True,
                "xFrameOptions": True,
                "xContentTypeOptions": True,
                "referrerPolicy": True,
                "score": 95
            },
            "rateLimiting": {
                "enabled": True,
                "requestsPerMinute": 1000,
                "blockedRequests": 12,
                "topBlockedIPs": ["192.168.1.100", "10.0.0.50"]
            },
            "lastSecurityScan": datetime.now() - timedelta(hours=6),
            "vulnerabilities": []
        },
        "backups": {
            "lastBackup": datetime.now() - timedelta(hours=2),
            "nextScheduledBackup": datetime.now() + timedelta(hours=22),
            "backupSize": 1073741824,  # 1GB
            "backupLocation": "/backups",
            "status": "success",
            "retentionDays": 30,
            "availableBackups": [
                {
                    "filename": "backup_2024_01_15_02_00.sql",
                    "date": datetime.now() - timedelta(hours=2),
                    "size": 1073741824,
                    "type": "full",
                    "verified": True
                },
                {
                    "filename": "backup_2024_01_14_02_00.sql", 
                    "date": datetime.now() - timedelta(days=1, hours=2),
                    "size": 1048576000,
                    "type": "full",
                    "verified": True
                }
            ]
        },
        "alerts": [],
        "lastUpdated": datetime.now()
    }

@router.get("/system/health")
async def get_system_health(current_user: models.User = Depends(get_current_user)):
    """Get overall system health status"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    return get_mock_system_health()

@router.get("/services/status")
async def get_service_status(current_user: models.User = Depends(get_current_user)):
    """Get status of all Docker services"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    return get_mock_system_health()["services"]

@router.post("/services/manage")
async def manage_service(
    action_data: Dict[str, Any],
    current_user: models.User = Depends(get_current_user)
):
    """Manage Docker services (restart, stop, start)"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    service = action_data.get("service")
    action = action_data.get("action")
    
    # Mock implementation - replace with actual Docker commands
    return {"message": f"Service {service} {action} completed successfully"}

@router.get("/services/{service_name}/logs")
async def get_service_logs(
    service_name: str,
    lines: int = 100,
    current_user: models.User = Depends(get_current_user)
):
    """Get logs for a specific service"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    # Mock log entries
    mock_logs = []
    for i in range(min(lines, 50)):
        mock_logs.append({
            "timestamp": datetime.now() - timedelta(minutes=i),
            "service": service_name,
            "level": "info" if i % 4 != 0 else "warning" if i % 8 != 0 else "error",
            "message": f"Sample log message {i} for {service_name}",
            "metadata": {},
            "traceId": f"trace-{i:04d}"
        })
    
    return mock_logs

@router.post("/logs/search")
async def search_logs(
    filter_data: Dict[str, Any],
    current_user: models.User = Depends(get_current_user)
):
    """Search and filter system logs"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    # Mock implementation
    return []

@router.get("/performance/metrics")
async def get_performance_metrics(
    timeRange: str = "1h",
    current_user: models.User = Depends(get_current_user)
):
    """Get system performance metrics"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    # Mock performance metrics
    return [
        {
            "name": "response_time",
            "value": 125.5,
            "unit": "ms",
            "trend": "stable",
            "threshold": 200,
            "chartData": [
                {"timestamp": datetime.now() - timedelta(minutes=i), "value": 120 + (i % 20)}
                for i in range(60, 0, -1)
            ]
        },
        {
            "name": "throughput",
            "value": 450.2,
            "unit": "req/s",
            "trend": "up",
            "threshold": 500,
            "chartData": [
                {"timestamp": datetime.now() - timedelta(minutes=i), "value": 400 + (i % 100)}
                for i in range(60, 0, -1)
            ]
        }
    ]

@router.get("/database/status")
async def get_database_status(
    current_user: models.User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get database status and metrics"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    try:
        # Test database connection
        db.execute(text("SELECT 1"))
        
        # Mock database status
        return {
            "connectionPool": {
                "active": 5,
                "idle": 15,
                "total": 20
            },
            "queryPerformance": {
                "averageResponseTime": 45.2,
                "slowQueries": 3,
                "totalQueries": 15420
            },
            "storage": {
                "size": 2147483648,  # 2GB
                "freeSpace": 8589934592,  # 8GB
                "tableCount": 25
            },
            "replication": {
                "status": "healthy",
                "lag": 0
            }
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Database error: {str(e)}")

@router.get("/redis/status")
async def get_redis_status(current_user: models.User = Depends(get_current_user)):
    """Get Redis cache status and metrics"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    # Mock Redis status
    return {
        "memory": {
            "used": 134217728,  # 128MB
            "peak": 268435456,  # 256MB
            "fragmentation": 15.2
        },
        "performance": {
            "hitRate": 94.5,
            "missRate": 5.5,
            "operationsPerSecond": 1250
        },
        "connections": {
            "connected": 8,
            "blocked": 0,
            "rejected": 0
        },
        "keyspace": {
            "totalKeys": 1542,
            "expiredKeys": 245,
            "evictedKeys": 12
        }
    }

@router.get("/backups/status")
async def get_backup_status(current_user: models.User = Depends(get_current_user)):
    """Get backup status and available backups"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    return get_mock_system_health()["backups"]["availableBackups"]

@router.post("/backups/create")
async def create_manual_backup(
    backup_request: Dict[str, Any],
    current_user: models.User = Depends(get_current_user)
):
    """Create a manual backup"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    # Mock backup creation
    backup_id = f"backup_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
    return {
        "backupId": backup_id,
        "message": "Backup created successfully"
    }

@router.get("/sessions")
async def get_active_sessions(current_user: models.User = Depends(get_current_user)):
    """Get active user sessions"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    # Mock active sessions
    return [
        {
            "id": "session-1",
            "userId": "user-1",
            "username": "admin",
            "role": "Owner",
            "ipAddress": "192.168.1.100",
            "userAgent": "Mozilla/5.0...",
            "loginTime": datetime.now() - timedelta(hours=2),
            "lastActivity": datetime.now() - timedelta(minutes=5),
            "isActive": True,
            "location": "Local Network"
        }
    ]

@router.get("/alerts")
async def get_system_alerts(current_user: models.User = Depends(get_current_user)):
    """Get system alerts"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    return []

@router.get("/ssl/status")
async def get_ssl_certificate_status(current_user: models.User = Depends(get_current_user)):
    """Get SSL certificate status"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    return get_mock_system_health()["security"]["sslCertificate"]

@router.get("/security/status")
async def get_security_status(current_user: models.User = Depends(get_current_user)):
    """Get security monitoring status"""
    if not current_user.role in ['Owner', 'Manager', 'Admin']:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    return get_mock_system_health()["security"]

File: frontend/src/types/systemAdmin.ts
export interface SystemHealth {
  overall: HealthStatus;
  services: ServiceHealth[];
  resources: ResourceUsage;
  security: SecurityStatus;
  backups: BackupHealth;
  alerts: SystemAlert[];
  lastUpdated: Date;
}

export interface HealthStatus {
  status: 'healthy' | 'warning' | 'critical';
  score: number; // 0-100
  message: string;
}

export interface ServiceHealth {
  name: string;
  status: 'healthy' | 'unhealthy' | 'starting' | 'stopped';
  uptime: string;
  cpu: number;
  memory: number;
  lastRestart: Date;
  actions: ServiceAction[];
  logs?: LogEntry[];
}

export interface ServiceAction {
  label: string;
  action: 'restart' | 'stop' | 'logs' | 'config';
  dangerous: boolean;
}

export interface ResourceUsage {
  cpu: {
    current: number;
    average: number;
    trend: 'up' | 'down' | 'stable';
  };
  memory: {
    used: number;
    total: number;
    percentage: number;
    trend: 'up' | 'down' | 'stable';
  };
  disk: {
    used: number;
    total: number;
    percentage: number;
    trend: 'up' | 'down' | 'stable';
  };
  network: {
    inbound: number;
    outbound: number;
  };
}

export interface SecurityStatus {
  sslCertificate: SSLCertificate;
  securityHeaders: SecurityHeaders;
  rateLimiting: RateLimitingStats;
  lastSecurityScan: Date;
  vulnerabilities: SecurityVulnerability[];
}

export interface SSLCertificate {
  domain: string;
  issuer: string;
  validFrom: Date;
  validTo: Date;
  daysUntilExpiry: number;
  status: 'valid' | 'expiring' | 'expired' | 'invalid';
  autoRenewal: boolean;
}

export interface SecurityHeaders {
  hsts: boolean;
  csp: boolean;
  xFrameOptions: boolean;
  xContentTypeOptions: boolean;
  referrerPolicy: boolean;
  score: number;
}

export interface RateLimitingStats {
  enabled: boolean;
  requestsPerMinute: number;
  blockedRequests: number;
  topBlockedIPs: string[];
}

export interface SecurityVulnerability {
  id: string;
  severity: 'low' | 'medium' | 'high' | 'critical';
  title: string;
  description: string;
  affectedService: string;
  discoveredAt: Date;
  status: 'open' | 'acknowledged' | 'fixed';
}

export interface BackupHealth {
  lastBackup: Date;
  nextScheduledBackup: Date;
  backupSize: number;
  backupLocation: string;
  status: 'success' | 'failed' | 'in_progress';
  retentionDays: number;
  availableBackups: BackupFile[];
}

export interface BackupFile {
  filename: string;
  date: Date;
  size: number;
  type: 'full' | 'incremental';
  verified: boolean;
}

export interface SystemAlert {
  id: string;
  type: 'info' | 'warning' | 'error' | 'critical';
  title: string;
  description: string;
  source: string;
  timestamp: Date;
  acknowledged: boolean;
  resolvedAt?: Date;
  actions: AlertAction[];
}

export interface AlertAction {
  label: string;
  action: string;
  dangerous: boolean;
}

export interface LogEntry {
  timestamp: Date;
  service: string;
  level: 'debug' | 'info' | 'warning' | 'error' | 'critical';
  message: string;
  metadata: Record<string, any>;
  traceId?: string;
}

export interface LogFilter {
  services: string[];
  levels: string[];
  dateRange: [Date, Date];
  searchQuery: string;
  limit: number;
}

export interface PerformanceMetric {
  name: string;
  value: number;
  unit: string;
  trend: 'up' | 'down' | 'stable';
  threshold: number;
  chartData: ChartDataPoint[];
}

export interface ChartDataPoint {
  timestamp: Date;
  value: number;
  label?: string;
}

export interface DatabaseStatus {
  connectionPool: {
    active: number;
    idle: number;
    total: number;
  };
  queryPerformance: {
    averageResponseTime: number;
    slowQueries: number;
    totalQueries: number;
  };
  storage: {
    size: number;
    freeSpace: number;
    tableCount: number;
  };
  replication: {
    status: 'healthy' | 'lagging' | 'failed';
    lag: number;
  };
}

export interface RedisStatus {
  memory: {
    used: number;
    peak: number;
    fragmentation: number;
  };
  performance: {
    hitRate: number;
    missRate: number;
    operationsPerSecond: number;
  };
  connections: {
    connected: number;
    blocked: number;
    rejected: number;
  };
  keyspace: {
    totalKeys: number;
    expiredKeys: number;
    evictedKeys: number;
  };
}

export interface FeatureFlag {
  name: string;
  description: string;
  enabled: boolean;
  rolloutPercentage: number;
  conditions: FeatureFlagCondition[];
  lastModified: Date;
  modifiedBy: string;
}

export interface FeatureFlagCondition {
  type: 'user_role' | 'user_id' | 'percentage' | 'date_range';
  operator: 'equals' | 'contains' | 'greater_than' | 'less_than';
  value: string | number | Date;
}

export interface EnvironmentVariable {
  key: string;
  value: string;
  sensitive: boolean;
  description?: string;
  lastModified: Date;
  modifiedBy: string;
}

export interface UserSession {
  id: string;
  userId: string;
  username: string;
  role: string;
  ipAddress: string;
  userAgent: string;
  loginTime: Date;
  lastActivity: Date;
  isActive: boolean;
  location?: string;
}

export interface SystemConfiguration {
  environmentVariables: EnvironmentVariable[];
  featureFlags: FeatureFlag[];
  systemSettings: Record<string, any>;
}

export interface ManualBackupRequest {
  type: 'full' | 'incremental';
  includeImages: boolean;
  includeConfigs: boolean;
  description?: string;
}

export interface RestoreRequest {
  backupFile: string;
  restoreType: 'full' | 'selective';
  selectedTables?: string[];
  confirmDataLoss: boolean;
}

export interface ServiceManagementAction {
  service: string;
  action: 'restart' | 'stop' | 'start';
  force?: boolean;
}